### PR TITLE
Add option to disable response verification

### PR DESCRIPTION
**Pull request by kevinhuangboku**
- Adding option to disable response signature verification
- Added tests
- Update readme

### CODE REVIEW

File: client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java [hunk 1]
```diff
@@ -342,14 +353,35 @@ private AuthorizationHeader getSignedAuthorizationHeader() {
...
+            if (respAuthHeaders.length != 1) {
+                throw new BokuAPIClientException(
+                        "Got " + httpResponse.getStatusLine() + " with " + respAuthHeaders.length + " " + AuthorizationHeader.RESPONSE_HEADER + " headers, expected 1!",
+                        apiClientResponse
+                );
+            }
+
+            verifySignature(httpResponse, apiClientResponse, respAuthHeaders);
+        } else {
+            if(respAuthHeaders.length == 0){
```
Comment (gmillaire) [inline client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java:369]: Another suggestion, as I don't normally like early returns like that.
```suggestion
            if(respAuthHeaders.length == 1){
                verifySignature(httpResponse, apiClientResponse, respAuthHeaders);
            } else if (respAuthHeaders.length > 1) {
                throw
            }
```

**Comment by gmillaire** [comment]
I added a suggestion, otherwise changes look good!

File: client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java [hunk 2]
```diff
@@ -257,7 +268,7 @@ private BokuAPIClientResponse executeAndReturnAPIResponse() throws IOException,
 
             // Verify signature on the response
             if (authHeader != null) {
-                verifyResponseSignature(httpResponse, apiClientResponse);
+                verifyResponseSignature(httpResponse, apiClientResponse, this.requireSignedResponse);
```
Comment (hokubokupoku) [inline client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java:271]: We want the logic to be,
 - If requireSignedResponse is true, require an X-SignedResponse header
 - If there is a X-SignedResponse header, verify it regardless

The old logic of checking for a request Authorization header to decide how to behave is now obsolete.

What I would do:
 - Get rid of the `authHeader != null` check
 - Move `Header[] respAuthHeaders = httpResponse.getHeaders(AuthorizationHeader.RESPONSE_HEADER)` out to this method.
 - If `requireSignedResponse` and there is no response header, produce the error from here.
 - If there is a response header, pass it into `verifyResponseSignature` - by the time that method is called, we know we require a valid signature, so it's either going to pass or throw an error.

Additionally you need to fix the flag setting behavior but I'll comment on that directly.

File: client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java [hunk 3]
```diff
@@ -161,6 +161,7 @@ public RequestBuilder delete(String url) {
         private final HttpUriRequest request;
         private AuthorizationHeader authHeader;
         private String entityString;
+        private boolean requireSignedResponse = true;
```
Comment (hokubokupoku) [inline client/src/main/java/com/boku/auth/http/client/BokuAPIClient.java:164]: This should default to false, and then get set to true by `withAuthorization`.

(Or, if you'd rather not make the options order-sensitive, you could have this be null, and then handle the defaulting behavior inside execute(), but I think it's simpler to just do it here.)

File: integration-tests/src/test/java/com/boku/auth/http/it/ClientTest.java [hunk 4]
```diff
@@ -129,6 +129,17 @@ public void testNoSignatureReturned() throws IOException {
...
 
+    @Test
+    public void testNoResponseSignatureRequired() throws IOException {
+        env.server.addServlet("/no-signature", Servlets.noAuth(new Servlets.PingHandler()));
+
+        env.client
+                .get(url("/no-signature"))
+                .withAuthorization(authorization())
+                .withOptionRequireSignedResponse(false)
+                .execute();
```
Comment (hokubokupoku) [inline integration-tests/src/test/java/com/boku/auth/http/it/ClientTest.java:140]: Lets assert on the response body, just to make sure that this worked.

**Comment by kevinhuangboku** [comment]
Followed documentation and merged these changes directly to master, closing this PR.

**Pull request closed without merging**